modal/_container_io_manager.py

```diff
@@ -43,12 +44,119 @@ class Sentinel:
 
 
 @dataclass
-class LocalInput:
-    input_id: str
-    function_call_id: str
-    method_name: str
-    args: Any
-    kwargs: Any
+class FinalizedFunction:
+    callable: Callable[..., Any]
+    is_async: bool
+    is_generator: bool
+    data_format: int  # api_pb2.DataFormat
+
+
+class IOContext:
+    input_ids: List[str]
+    function_call_ids: List[str]
+    finalized_function: FinalizedFunction
+
+    def __init__(
+        self,
+        input_ids: List[str],
+        function_call_ids: List[str],
+        finalized_function: FinalizedFunction,
+        deserialized_args: List,
+        is_batched: bool,
+    ):
+        self.input_ids = input_ids
+        self.function_call_ids = function_call_ids
+        self.finalized_function = finalized_function
+        self._deserialized_args = deserialized_args
+        self._is_batched = is_batched
+
+    @classmethod
+    async def create(
+        cls,
+        client: _Client,
+        finalized_functions: Dict[str, FinalizedFunction],
+        inputs: List[Tuple[str, str, api_pb2.FunctionInput]],
+        is_batched: bool,
+    ) -> "IOContext":
+        assert len(inputs) >= 1 if is_batched else len(inputs) == 1
+        input_ids, function_call_ids, inputs = zip(*inputs)
+
+        async def _populate_input_blobs(client: _Client, input: api_pb2.FunctionInput):
+            # If we got a pointer to a blob, download it from S3.
+            if input.WhichOneof("args_oneof") == "args_blob_id":
+                args = await blob_download(input.args_blob_id, client)
+                # Mutating
+                input.ClearField("args_blob_id")
+                input.args = args
+
+            return input
+
+        inputs = await asyncio.gather(*[_populate_input_blobs(client, input) for input in inputs])
+        # check every input in batch executes the same function
+        method_name = inputs[0].method_name
+        assert all(method_name == input.method_name for input in inputs)
+        finalized_function = finalized_functions[method_name]
+        # TODO(cathy) Performance decrease if we deserialize inputs later
+        deserialized_args = [deserialize(input.args, client) if input.args else ((), {}) for input in inputs]
+        return cls(input_ids, function_call_ids, finalized_function, deserialized_args, is_batched)
+
+    def _args_and_kwargs(self) -> Tuple[Tuple[Any, ...], Dict[str, List]]:
+        if not self._is_batched:
+            return self._deserialized_args[0]
+
```

Comment:
invalid error if generator?

Reply:
Removed this and function default arguments check because this will be checked earlier (in functions.py `from_args`)